### PR TITLE
serialize to bytebufer and deserialization from byte[] to avoid intermediate String instantiation

### PR DESCRIPTION
**Pull request by moderakh**
As per jackson json serialization/desrialization library best practices guideline:
* when deserializing json it is better to do deserialization directly form byte[]
* when serializing to json, it is better to serialize to OutputStream

The purpose of this PR is to set the foundation of the following:
* on the write path change serialization to directly serialize to ByteBuffer (prior to this PR serialization was happening in two steps from POJO -> String and from String -> byte[])
* on the read path change deserialization to directly read from byte[].

I am planning to do more cleanup on read/write handling and byte[]/ByteBuffer handling after this PR.
More PRs coming to fully move to new model in other places.

### CODE REVIEW

**Comment by kirankumarkolli** [comment]
This will sets us up right design/architecture wise. Thanks @moderakh.

**Comment by kirankumarkolli** [comment]
QQ: from when string intermediate materialization present?

---
In reply to: [582319430](https://github.com/Azure/azure-sdk-for-java/pull/7983#issuecomment-582319430) [](ancestors = 582319430)

**Comment by kirankumarkolli** [comment]
>     static String toJsonString(Object cosmosItem, ObjectMapper objectMapper) {

Who are the consumers? Isn;t it best to remove it?

---
Refers to: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosItemProperties.java:79 in 46618b6. [](commit_id = 46618b6382c1dc71a851eaf802b613f051239eb6, deletion_comment = False)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java [hunk 1]
```diff
@@ -529,4 +535,37 @@ public ValueHolder(V v) {
...
+            return getSimpleObjectMapper().readValue(item, itemClassType);
+        } catch (IOException e) {
+            throw new IllegalStateException("Failed to get POJO.", e);
+        }
+    }
+
+    public static ByteBuffer serializeJsonToByteBuffer(ObjectMapper objectMapper, Object object) {
+        try {
+            ByteBufferOutputStream byteBufferOutputStream = new ByteBufferOutputStream();
+            objectMapper.writeValue(byteBufferOutputStream, object);
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java]: Will it result in re-allocations (dynamic size)?

Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Utils.java]: yes

**Comment by moderakh** [comment]
> QQ: from when string intermediate materialization present?
@kirankumarkolli 
on the readpath: there won't be any materialization to string. deserialization to user type happen directly from byte[]
on writepath: there won't be any intermediate string. we directly serialize user custom type to bytebuf

**Comment by moderakh** [comment]
> > ```
> > static String toJsonString(Object cosmosItem, ObjectMapper objectMapper) {
> > ```
> 
> Who are the consumers? Isn;t it best to remove it?
> 
> Refers to: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosItemProperties.java:79 in [46618b6](https://github.com/Azure/azure-sdk-for-java/commit/46618b6382c1dc71a851eaf802b613f051239eb6). [](commit_id = [46618b6](https://github.com/Azure/azure-sdk-for-java/commit/46618b6382c1dc71a851eaf802b613f051239eb6), deletion_comment = False)

that's not used and removed. PR was in intermediate draft mode state.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosItemProperties.java [hunk 2]
```diff
@@ -57,21 +76,23 @@ static Document fromObject(Object cosmosItem) {
...
-            } catch (IOException e) {
-                throw new IllegalArgumentException("Can't serialize the object into the json string", e);
+                return ((Document) cosmosItem).serializeJsonToByteBuffer();
             }
+
+            return Utils.serializeJsonToByteBuffer(objectMapper, cosmosItem);
         }
     }
 
+    static List<CosmosItemProperties> getFromV2Results(List<Document> results) {
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosItemProperties.java]: I don't think this function is needed anymore, We can remove this.

Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosItemProperties.java]: removed. thanks

**Comment by moderakh** [comment]
cherry picked PR is merged.

**Pull request closed without merging**